vendors/nordic/boards/nrf52840-dk/ports/ota_pal_for_aws/ota_pal.c

```diff
@@ -0,0 +1,734 @@
+
+/*-----------------------------------------------------------*/
+
+OtaPalMainStatus_t prvErasePages( size_t xFrom,
+                                  size_t xTo )
+{
+    size_t ulErasedAddress = ( size_t ) xFrom;
+    OtaPalMainStatus_t xStatus = OtaPalSuccess;
+    ret_code_t xErrCode = NRF_SUCCESS;
+
+    while( xStatus == OtaPalSuccess && ulErasedAddress < ( size_t ) xTo )
+    {
+        size_t ulErasedPage = ( size_t ) ( ulErasedAddress / NRF_FICR->CODEPAGESIZE );
+        ulErasedAddress += NRF_FICR->CODEPAGESIZE;
+        xEventGroupClearBits( xFlashEventGrp, otapalFLASH_SUCCESS | otapalFLASH_FAILURE );
+        xErrCode = sd_flash_page_erase( ulErasedPage );
+
+        if( xErrCode != NRF_SUCCESS )
+        {
+            xStatus = OtaPalRxFileCreateFailed;
+        }
+        else
+        {
+            /**
+             * If soft device is enabled, the result of sd_flash_page_erase() is posted via event, otherwise sd_flash_page_erase()
+             * is a blocking operation.
+             */
+            if( nrf_sdh_is_enabled() )
+            {
+                EventBits_t xFlashResult = xEventGroupWaitBits( xFlashEventGrp, otapalFLASH_SUCCESS | otapalFLASH_FAILURE, pdTRUE, pdFALSE, portMAX_DELAY );
+
+                if( xFlashResult & otapalFLASH_FAILURE )
+                {
+                    xStatus = OtaPalRxFileCreateFailed;
+                }
+            }
+        }
+    }
+
+    return xStatus;
+}
+
+OtaPalStatus_t otaPal_CreateFileForRx( OtaFileContext_t * const pFileContext )
+{
+    ret_code_t xErrCode;
+
+    pFileContext->pFile = prvNextFreeFileHandle;
+    OtaPalMainStatus_t xStatus = OtaPalSuccess;
+
+    /* Check that the second bank size is big enough to contain the new firmware */
+    if( pFileContext->fileSize > otapalSECOND_BANK_SIZE )
+    {
+        xStatus = OtaPalOutOfMemory;
+    }
+
+    if( xStatus == OtaPalSuccess )
+    {
+        /* Create the event group for flash events */
+        xFlashEventGrp = xEventGroupCreate();
+
+        if( xFlashEventGrp == NULL )
+        {
+            LogInfo( ( "No memory for creation of event group " ) );
+            xStatus = OtaPalOutOfMemory;
+        }
+    }
+
+    if( xStatus == OtaPalSuccess )
+    {
+        /* Erase the required memory */
+        LogInfo( ( "Erasing the flash memory " ) );
+        xStatus = prvErasePages( otapalSECOND_BANK_START, otapalSECOND_BANK_END );
+
+        if( xStatus != OtaPalSuccess )
+        {
+            LogInfo( ( "Erasing the flash memory failed with error_code %d", xStatus ) );
+        }
+    }
+
+    return OTA_PAL_COMBINE_ERR( xStatus, 0 );
+}
+/*-----------------------------------------------------------*/
+
+OtaPalStatus_t otaPal_Abort( OtaFileContext_t * const pFileContext )
+{
+    pFileContext->pFile = ( int32_t ) NULL;
+
+    /* Delete the event group */
+    if( xFlashEventGrp != NULL )
+    {
+        vEventGroupDelete( xFlashEventGrp );
+        xFlashEventGrp = NULL;
+    }
+
+    return OtaPalSuccess;
```

Comment:
The return type is OtaPalStatus_t. Please use OTA_PAL_COMBINE_ERR to shift the error code to the correct position.